apps/aecore/lib/aecore/structures/spend_tx.ex

```diff
@@ -78,29 +79,36 @@ defmodule Aecore.Structures.SpendTx do
   """
   @spec process_chainstate!(
           SpendTx.t(),
-          binary(),
+          Wallet.pubkey(),
           non_neg_integer(),
           non_neg_integer(),
           non_neg_integer(),
-          ChainState.account(),
+          AccountStateTree.tree(),
           tx_type_state()
-        ) :: {ChainState.accounts(), tx_type_state()}
-  def process_chainstate!(%SpendTx{} = tx, sender, fee, nonce, block_height, accounts, %{}) do
-    sender_account_state = Map.get(accounts, sender, Account.empty())
+        ) :: {AccountStateTree.tree(), tx_type_state()} | {:error, String.t()}
+  def process_chainstate!(%SpendTx{} = tx, sender, fee, nonce, _block_height, accounts, %{}) do
+    process_chainstate!(tx, sender, fee, nonce, accounts, %{})
+  end
 
-    preprocess_check!(tx, sender, sender_account_state, fee, block_height, %{})
+  defp process_chainstate!(%SpendTx{} = tx, sender, fee, nonce, accounts, %{}) do
+    sender_account_state = Account.get_account_state(accounts, sender)
 
-    new_sender_account_state =
-      sender_account_state
-      |> deduct_fee(fee)
-      |> Account.transaction_out(tx.amount * -1, nonce)
+    case preprocess_check!(tx, sender_account_state, fee) do
```

Comment:
Shouldn't you call the full function of preprocess_check!/7 here ?

Reply:
We think that is more readable and clean. You don't have to pass unused arguments.

Reply:
I would adjust this, to keep it the same as the transaction interface